moz-webgpu-cts/src/shared.rs

```diff
@@ -672,28 +634,113 @@ impl Display for MetadataTestPathError<'_> {
+    ) -> Result<(TestScope, &'a Utf8Path), std::path::StripPrefixError> {
+        let strip_prefix =
+            |prefix, scope| path.strip_prefix(prefix).map(|stripped| (scope, stripped));
+        match self {
+            Browser::Firefox => {
+                strip_prefix(SCOPE_DIR_FX_MOZILLA_STR, FirefoxTestScope::Mozilla.into()).or_else(
+                    |_| strip_prefix(SCOPE_DIR_FX_UPSTREAM_STR, FirefoxTestScope::Upstream.into()),
+                )
+            }
+            Browser::Servo => {
+                strip_prefix(SCOPE_DIR_SERVO_WEBGPU_STR, ServoTestScope::WebGpu.into())
+            }
+        }
+    }
+}
+
+/// Symbolically represents a file root from which tests and metadata are based. Scopes are based
+/// on a specific [`Browser`].
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub(crate) enum TestScope {
+    Firefox(FirefoxTestScope),
+    Servo(ServoTestScope),
+}
+
+impl TestScope {
+    /// NOTE: Keep this implementation in sync with [`Browser::strip_scope_url_prefix`].
+    fn url_prefix(&self) -> &str {
+        match self {
+            TestScope::Firefox(scope) => match scope {
+                FirefoxTestScope::Upstream => "",
+                FirefoxTestScope::Mozilla => "_mozilla/",
+            },
+            TestScope::Servo(ServoTestScope::WebGpu) => "_webgpu/",
+        }
+    }
+
+    /// NOTE: Keep this implementation in sync with [`Browser::strip_scope_metadata_parent_path`].
+    fn metadata_parent_path_components(&self) -> impl Iterator<Item = &str> + Clone {
+        match self {
+            TestScope::Firefox(scope) => match scope {
+                FirefoxTestScope::Upstream => SCOPE_DIR_FX_UPSTREAM_COMPONENTS,
+                FirefoxTestScope::Mozilla => SCOPE_DIR_FX_MOZILLA_COMPONENTS,
+            },
+            TestScope::Servo(ServoTestScope::WebGpu) => SCOPE_DIR_SERVO_WEBGPU_COMPONENTS,
+        }
+        .iter()
+        .cloned()
+    }
+}
+
+/// Subset of [`TestScope`] for [`Browser::Firefox`].
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub(crate) enum FirefoxTestScope {
     /// A public test available at some point in the history of [WPT upstream]. Note that while
     /// a test may be public, metadata associated with it is in a private location.
     ///
     /// [WPT upstream]: https://github.com/web-platform-tests/wpt
-    Public,
-    /// A private test specific to browser.
-    Private,
+    Upstream,
+    /// A private test specific to Firefox.
+    Mozilla,
+}
+
+impl From<FirefoxTestScope> for TestScope {
+    fn from(value: FirefoxTestScope) -> Self {
+        Self::Firefox(value)
+    }
 }
 
-/// Symbolically represents a file root from which tests and metadata are based.
+/// Subset of [`TestScope`] for [`Browser::Servo`].
 #[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
-pub(crate) struct TestScope {
-    browser: Browser,
-    visibility: TestVisibility,
+pub(crate) enum ServoTestScope {
+    WebGpu,
```

Comment:
From @sagudev on the commit ([original link](https://github.com/ErichDonGubler/moz-webgpu-cts/commit/d04e532c02cf13744de69183bcce05973f752dd2#r141916888)):

> Maybe add some rustdoc comment that would describe variant. Something about that this is WebGPU CTS directory.

Reply:
@sagudev: How about daa3f736?

Reply:
This is nitpicking but `source` sounds wired to me in this context, maybe `Servo tree` or `Servo repo` would be more appropriate?

Disclaimer: I am NOT a native speaker.

Reply:
I consider the terms "source code", "tree" (short for "source tree"), and "repo" to be largely interchangeable. 🙂 Maybe "Servo's source tree" works, then?

Reply:
Pushed my suggestion up as 0546b5f70dc889e4f7d466edf5da6405bee9cc20.